barretenberg/cpp/src/barretenberg/dsl/acir_format/ec_operations.cpp

```diff
@@ -13,19 +14,126 @@
 
-    auto input1_point = to_grumpkin_point(
-        input.input1_x, input.input1_y, input.input1_infinite, has_valid_witness_assignments, builder);
-    auto input2_point = to_grumpkin_point(
-        input.input2_x, input.input2_y, input.input2_infinite, has_valid_witness_assignments, builder);
+    // Check if operands are the 'same' (same witness or same constant value).
+    bool x_match = false;
+    if (!input.input1_x.is_constant && !input.input2_x.is_constant) {
+        x_match = (input.input1_x.index == input.input2_x.index);
+    } else {
+        if (input.input1_x.is_constant && input.input2_x.is_constant) {
+            x_match = (input.input1_x.value == input.input2_x.value);
+        }
+    }
+    bool y_match = false;
+    if (!input.input1_y.is_constant && !input.input2_y.is_constant) {
+        y_match = (input.input1_y.index == input.input2_y.index);
+    } else {
+        if (input.input1_y.is_constant && input.input2_y.is_constant) {
+            y_match = (input.input1_y.value == input.input2_y.value);
+        }
+    }
+
+    cycle_group_ct result;
+    // If operands are the same, we double.
+    // Note that the doubling function handles the infinity case
+    if (x_match && y_match) {
+        cycle_group_ct input1_point;
+
+        // When there are no valid witness assignements, we need to define dummy values that will
+        // satisfy the doubling constraints, which we can do easily when the inputs are witness.
+        // If the is_infinity is a witness, we can simply set it to 1
+        // Or, if the coordinates are witness, we can simply set them to a valid point on the curve (G1)
+        if (!input.input1_infinite.is_constant || (!input.input1_x.is_constant && !input.input1_y.is_constant)) {
+            input1_point = to_grumpkin_point(
+                input.input1_x, input.input1_y, input.input1_infinite, has_valid_witness_assignments, true, builder);
+        } else {
+            // If not, the coordinates are mixed constant/witness, and we generate witness so that the point is using
+            // only witnesses.
+            input1_point = to_witness_grumpkin_point(
+                input.input1_x, input.input1_y, input.input1_infinite, has_valid_witness_assignments, true, builder);
+        }
+        result = input1_point.dbl();
+    } else {
+        // Regular addition
+        // if one point is (constant) zero, we simply return the other point.
+        if (input.input2_infinite.is_constant && input.input1_infinite.is_constant) {
+            if (get_value(input.input1_infinite, builder) == 1) {
+                // input1 is infinity, so we can just return input2
+                result = to_grumpkin_point(input.input2_x,
+                                           input.input2_y,
+                                           input.input2_infinite,
+                                           has_valid_witness_assignments,
+                                           false,
```

Comment:
These two ifs are equivalent, but you create the point with 'use_g1' being set to different values

Reply:
Any valid point would do here, it is just for consistency, as I use G1 for point1 and G2=2G1 for point2 elsewhere in the file.
But I don't mind changing it.